lib/upsert/merge_function/PG_Connection.rb

```diff
@@ -6,7 +6,67 @@ class MergeFunction
     class PG_Connection < MergeFunction
       include Postgresql
 
+      attr_reader :quoted_setter_names
+      attr_reader :quoted_selector_names
+
+      def initialize(controller, *args)
+        super
+        @quoted_setter_names = setter_keys.map { |k| connection.quote_ident k }
+        @quoted_selector_names = selector_keys.map { |k| connection.quote_ident k }
+      end
+
       def execute(row)
+        use_pg_upsert? ? pg_upsert(row) : pg_function(row)
+      end
+
+      def use_pg_upsert?
+        return false if disable_native?
+        return @use_upsert if defined?(@use_upsert)
+        @use_upsert = begin
+          version = controller.connection.execute("SHOW server_version").getvalue(0, 0)
+          version_number = version[0..2].split('.').join('').to_i
+
+          return @use_upsert = false if version_number < 95
+
+          schema_query = controller.connection.execute(%{
+              SELECT array_agg(column_name::text) FROM information_schema.constraint_column_usage
+              LEFT JOIN pg_catalog.pg_constraint ON constraint_name::text = conname::text
+              WHERE table_name = $1 AND conrelid = $1::regclass::oid AND contype = 'u'
+              GROUP BY table_catalog, table_name, constraint_name
+          }, [table_name])
+          type_map = PG::TypeMapByColumn.new([PG::TextDecoder::Array.new])
+          schema_query.type_map = type_map
+
+          matching_constraint = schema_query.values.any? do |row|
+            row.first.sort == selector_keys.sort
+          end
+
+          version_number >= 95 && matching_constraint
+        end
+      end
+
+      def pg_upsert(row)
```

Comment:
i think you should rename this to `pg_native`

(so the options would be `pg_function` or `pg_native`)